workers/unity/Packages/io.improbable.gdk.tools/.CodeGenTemplate/CodeGenerationLib/CodeWriter/Scopes/Abstract/ScopeBody.cs

```diff
@@ -0,0 +1,107 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Improbable.Gdk.CodeGeneration.CodeWriter.Scopes
+{
+    /// <summary>
+    /// A ScopeBlock with helpers for writing code found in method bodies.
+    /// </summary>
+    public abstract class ScopeBody : ScopeBlock
+    {
+        protected ScopeBody(string declaration, Action<ScopeBody> populate) : base(declaration)
+        {
+            populate(this);
+        }
+
+        protected ScopeBody(string declaration = null) : base(declaration)
+        {
+        }
+
+        public void WriteLine(string snippet)
+        {
+            Add(new Text(snippet));
+        }
+
+        public void WriteLine(Action<StringBuilder> populate)
+        {
+            var sb = new StringBuilder();
+            populate(sb);
+
+            Add(new Text(sb.ToString()));
+        }
+
+        public void WriteLine(IEnumerable<string> snippets)
+        {
+            if (snippets.Any())
+            {
+                Add(new TextList(snippets));
+            }
+        }
+
+        public void Initializer(string declaration, Func<IEnumerable<string>> populate)
+        {
+            Add(new InitializerCodeBlock(declaration, populate));
+        }
+
+        public void CustomScope(Action<CustomScopeBlock> populate)
+        {
+            Add(new CustomScopeBlock(populate));
+        }
+
+        public void CustomScope(string declaration, Action<CustomScopeBlock> populate)
+        {
+            Add(new CustomScopeBlock(declaration, populate));
+        }
+
+        public void Loop(string declaration, Action<LoopBlock> populate)
```

Comment:
What is the declaration in the `Loop` block?

Reply:
Is this a while loop?

Reply:
while/for/foreach

do-while hasn't really been considered yet but if we want thtat it can be done in a similar way to the if-else/try-catch blocks